packages/contractkit/src/wallets/ledger-wallet.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  ensureLeading0x,
+  isHexString,
+  normalizeAddressWith0x,
+  trimLeading0x,
+} from '@celo/utils/lib/address'
+import TransportNodeHid from '@ledgerhq/hw-transport-node-hid'
+import * as ethUtil from 'ethereumjs-util'
+import { EncodedTransaction, Tx } from 'web3-core'
+import { Address } from '../base'
+import { EIP712TypedData, generateTypedDataHash } from '../utils/sign-typed-data-utils'
+import { encodeTransaction, rlpEncodedTx } from '../utils/signing-utils'
+import { AppEth } from './ledger-types'
+import { Wallet } from './wallet'
+
+export const CELO_BASE_DERIVATION_PATH = "44'/52752'/0'/0"
+const ADDRESS_QTY = 5
+
+export class LedgerWallet implements Wallet {
+  // Account addresses are hex-encoded, lower case alphabets
+  private readonly addressDerivationPath = new Map<Address, string>()
+  private addressesRetrieved = false
+  private setupFinished = false
+  private appEth: AppEth | undefined
+
+  /**
+   * @param derivationPathIndexes number array of "address_index" for the base derivation path.
+   * Default: Array[0..9].
+   * Example: [3, 99, 53] will retrieve the derivation paths of
+   * [`${baseDerivationPath}/3`, `${baseDerivationPath}/99`, `${baseDerivationPath}/53`]
+   * @param baseDerivationPath base derivation path. Default: "44'/52752'/0'/0"
+   * @param devicePath path to the ledger device, otherwise will use the first one it finds
+   */
+  constructor(
+    readonly derivationPathIndexes: number[] = Array.from(Array(ADDRESS_QTY).keys()),
+    readonly baseDerivationPath: string = CELO_BASE_DERIVATION_PATH,
+    readonly devicePath: string = ''
+  ) {
+    const validDPs = derivationPathIndexes.reduce(
+      (valid, currentValue) => valid && Number.isInteger(currentValue) && currentValue >= 0,
+      true
+    )
+    if (!validDPs) {
+      throw new Error('ledger-wallet: Invalid address index')
+    }
+    this.setup()
```

Comment:
I have been told that this isn't an idiomatic pattern
Any async initialization should happen outside of the constructor

Reply:
agree

Reply:
Wasn't suppose to keep that on the commit (sorry)